File: 2. Recursion/Medium and Hard/Subsets.py
# Subsets (Leetcode)
# Leetcode Link : https://leetcode.com/problems/subsets/description/

# Time Complexity : O(2^n)
# Space Complexity : O(2^n)

class Solution:
    def subsets(self, nums: List[int]) -> List[List[int]]:
        ans = []
        
        def solve(nums, op):
            # If input is empty, add the output i.e. ds to the ans
            if len(nums) == 0:
                ans.append(op)
                return
            
            # Making i/p smaller by erasing 1st element, ignoring element at index '0' for output
            solve(nums[1:], op.copy())
            # Making i/p smaller by erasing 1st element, taking element at index '0' for output
            solve(nums[1:], op + [nums[0]])
        
        solve(nums, [])
        return ans

File: 11. Bit Manipulation/2. Interview Problems/Subsets.cpp
// Subsets (Leetcode)
// Leetcode Link : https://leetcode.com/problems/subsets/description/

// Time Complexity : O(2^n * n)
// Space Complexity : O(2^n * n)

class Solution {
public:
    vector<vector<int>> subsets(vector<int>& nums) {
        int n = nums.size();
        int noOfSubsets = (1 << n);
        vector<vector<int>> subsets;

        for(int i = 0; i < noOfSubsets; i++){
            vector<int> subset;
            // Iterate through each element of the input array
            for(int j = 0; j < n; j++){
                // Check if the jth bit of i is set
                if(i & 1 << j){
                    // If yes, include the jth element of nums in the current subset
                    subset.push_back(nums[j]);
                }
            }
            subsets.push_back(subset);
        }

        return subsets;
    }
};